### PR TITLE
Get SHDI for AOI by intersection using SQL

### PR DESCRIPTION
**Pull request by matthiasschaub**
### Description

 Get SHDI for AOI by intersection using SQL.

*Note: The development database needs to be rebuild for the changes to take effect.*

### Checklist
- [x] I have updated my branch to `main` (e.g. through `git rebase main`)
- [x] My code follows the [style guide](https://github.com/GIScience/ohsome-quality-analyst/blob/main/CONTRIBUTING.md#style-guide) and was checked with [pre-commit](https://github.com/GIScience/ohsome-quality-analyst/blob/main/CONTRIBUTING.md#pre-commit) before committing
- [x] I have commented my code
- [x] I have added sufficient unit and integration [tests](https://github.com/GIScience/ohsome-quality-analyst/blob/main/docs/development_setup.md#tests)
- [x] I have updated the [CHANGELOG.md](https://github.com/GIScience/ohsome-quality-analyst/blob/main/CHANGELOG.md)
- [x] Download and import SHDI during docker build for development image
- [x] Import SHDI into production database
- ~~[ ] Add spatial index~~
- [x] Add SHDI for development and test regions to the database
- [x] Create issue: Check possibility to update shapefile in database to SHDI v.5

### CODE REVIEW

File: workers/tests/integrationtests/test_geodatabase.py [hunk 1]
```diff
@@ -139,6 +139,29 @@ def test_map_fid_to_uid(self):
...
         self.assertEqual(result, "3")
 
+    def test_get_shdi_single_intersection(self):
+        """Input geometry intersects only with one (shdi) region."""
+        shdi = asyncio.run(db_client.get_shdi(self.feature.geometry))
+        self.assertIsInstance(shdi, float)
+        self.assertLessEqual(shdi, 1.0)
+
+    def test_get_shdi_multiple_intersections(self):
+        """Input geometry intersects only with one (shdi) region."""
```
Comment (matthiasschaub) [inline workers/tests/integrationtests/test_geodatabase.py:149]: Wrong docstring

File: docs/development_setup.md [hunk 2]
```diff
@@ -17,7 +17,7 @@ Please continue reading for more information on each one of those services. If r
 
 ## Database
 
-A database for development purposes is provided as Dockerfile. This database contains custom regions, regions for running tests and the GHS POP dataset for those regions. To build and run an already configured image run:
+A database for development purposes is provided as Dockerfile. This database contains custom regions, regions for running tests and dataset (SHDI and GHS-POP) for those regions. To build and run an already configured image run:
```
Comment (Hagellach37) [inline docs/development_setup.md:20]: small type, it should be "datasets"

File: workers/ohsome_quality_analyst/geodatabase/client.py [hunk 3]
```diff
@@ -255,3 +253,20 @@ async def map_fid_to_uid(dataset: str, feature_id: str, fid_field: str) -> str:
...
     return str(record[0])
+
+
+async def get_shdi(bpoly: Union[Polygon, MultiPolygon]) -> float:
+    """Get Subnational Human Development Index (SHDI) for a bounding polygon.
+
+    Get SHDI by intersecting the bounding polygon with sub-national regions provided by
+    the GlobalDataLab (GDL).
+
+    If intersection with multiple GDL regions occurs, return the sum of SHDI in
```
Comment (Hagellach37) [inline workers/ohsome_quality_analyst/geodatabase/client.py:264]: I think it's not the "sum in proportion of the intersection area" but should be the "weighted average (using area as the weight)"

**Review by Hagellach37 — COMMENTED** [review]
Hey @matthiasschaub ,
thanks for adding the SHDI to the database. 

Besides the comments I made already, I was wondering if we should use a spatial index on the shdi table. I think we don't have it so far. There should be an option to add a spatial index (e.g. gist) during the import or directly afterwards.

Let me know if you have further questions.

File: docs/development_setup.md (hunk 2 above)
Comment (matthiasschaub) [reply docs/development_setup.md:20]: Fixed typo

File: workers/ohsome_quality_analyst/geodatabase/client.py (hunk 3 above)
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/geodatabase/client.py:264]: Yes, using your formulation is much clearer. I corrected the docstring.

File: workers/ohsome_quality_analyst/geodatabase/select_shdi.sql [hunk 4]
```diff
@@ -0,0 +1,12 @@
+WITH bpoly AS (
+    SELECT
+        ST_Setsrid (public.ST_GeomFromGeoJSON ($1), 4326) AS geom
+)
+SELECT
+    SUM(ST_Area (ST_Intersection (shdi.geom, bpoly.geom)) / ST_Area (bpoly.geom) *
```
[the remark this answers is not in the document]
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/geodatabase/select_shdi.sql]: Good catch! I cast the geometry to geogprahy type now

**Comment by matthiasschaub** [comment]
> Hey @matthiasschaub , thanks for adding the SHDI to the database.
> 
> Besides the comments I made already, I was wondering if we should use a spatial index on the shdi table. I think we don't have it so far. There should be an option to add a spatial index (e.g. gist) during the import or directly afterwards.
> 
> Let me know if you have further questions.

Thanks for the review! I addressed all the inline comments.

I think a spatial index is not need for a table this small. PostGIS docs are stating:
> Once a spatial data table exceeds a few thousand rows, you will want to build an index to speed up spatial searches of the data (unless all your searches are based on attributes, in which case you'll want to build a normal index on the attribute fields).
~ https://postgis.net/docs/using_postgis_dbmanagement.html#build-indexes

File: workers/tests/integrationtests/test_geodatabase.py [hunk 5]
```diff
@@ -139,6 +139,30 @@ def test_map_fid_to_uid(self):
         )
         self.assertEqual(result, "3")
 
+    def test_get_shdi_single_intersection(self):
+        """Input geometry intersects only with one SHDI region."""
+        shdi = asyncio.run(db_client.get_shdi(self.feature.geometry))
+        self.assertIsInstance(shdi, float)
+        self.assertLessEqual(shdi, 1.0)
+
+    # Note: This test can only be executed if the whole SHDI is in the database.
```
Comment (Hagellach37) [inline workers/tests/integrationtests/test_geodatabase.py:148]: would you leave this here or rather remove?

**Review by Hagellach37 — APPROVED** [review]
looks good to me now. :) what about the unused test?

**Review by Hagellach37 — APPROVED** [review]

File: workers/tests/integrationtests/test_geodatabase.py (hunk 5 above)
Comment (matthiasschaub) [reply workers/tests/integrationtests/test_geodatabase.py:148]: I want to leave it to make clear that this behavior has been tested.

**Pull request merged**